### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by abarnes6**
With the help of some AI, the code is looking pretty spiffy. Has not been tested yet, will do that at some point.

### CODE REVIEW

**Review by BigBroccoli — COMMENTED** [review]
Looks good will approve once this is tested on the bot and proven to work

**Review by BigBroccoli — APPROVED** [review]
Looks good and works

**Review by submergedcastle — APPROVED** [review]
virus 👍

**Review by Kurosaki077 — APPROVED** [review]
No clue what's happening here but looks awesome.

**Review by peyton-otey — COMMENTED** [review]
Didn't understand this, here's a ChatGPT o1-preview output:

1. diagnostics.py: **Issue in update_diag Method**

    Problem: The method attempts to update an existing DiagnosticStatus in self.diagnostic_info.status by comparing hardware_id. However, the assignment status = info inside the loop only changes the local variable status and does not update the list.

    Impact: The diagnostic information will not be updated correctly, potentially causing outdated or incorrect status reports.

    Suggestion: Modify the loop to update the list item directly using its index. Here's how you can fix it:

```
def update_diag(self, info: DiagnosticStatus):
    try:
        for idx, status in enumerate(self.diagnostic_info.status):
            if status.hardware_id == info.hardware_id:
                self.diagnostic_info.status[idx] = info
                self.get_logger().debug(f'Updated diagnostic info for hardware_id: {info.hardware_id}')
                break
        else:
            self.diagnostic_info.status.append(info)
            self.get_logger().debug(f'Added new diagnostic info for hardware_id: {info.hardware_id}')
    except Exception as e:
        self.get_logger().error(f'Failed to update diagnostic info: {e}')
```

This code ensures that the status is updated correctly and that new hardware IDs are appended properly.




**Missing rclpy.shutdown() Call**

    Problem: In the main function, rclpy.shutdown() has been removed.

    Impact: Not calling rclpy.shutdown() can lead to improper shutdown of ROS nodes, which may cause resource leaks or other unexpected behavior.

    Suggestion: Reintroduce the rclpy.shutdown() call in the finally block to ensure that the node shuts down cleanly:

```
def main(args=None):
    rclpy.init(args=args)
    diagnostic_node = Diagnostics()
    try:
        rclpy.spin(diagnostic_node)
    except KeyboardInterrupt:
        diagnostic_node.get_logger().warning('Ctrl-C pressed, shutting down...')
    finally:
        rclpy.shutdown()
```


2. serial_bridge.py: **Issues in set_parameters Method**

    Problem 1: Exception handling is overly broad and may catch exceptions not related to serial communication.

    Impact: This can make debugging difficult and may mask other issues.

    Suggestion: Catch specific exceptions (e.g., SerialException) to make error handling more precise.

**Review by peyton-otey — APPROVED** [review]
Approved!

**Review by afb29 — APPROVED** [review]
LGTM

**Comment by abarnes6** [comment]
@peyton-otey I will update the PR with a couple of changes in response to that feedback

File: src/moondawg/moondawg/diagnostics.py [hunk 1]
```diff
@@ -15,7 +12,7 @@ def __init__(self):
 
         # create heartbeat
```
Comment (Reeree360) [inline src/moondawg/moondawg/diagnostics.py:13]: explain what this means and does

Comment (abarnes6) [reply src/moondawg/moondawg/diagnostics.py:13]: Diagnostics was kinda neglected in this PR, a lot can be improved for when you and Tony tackle it for improvement

**Review by Reeree360 — APPROVED** [review]
I know I left a lot of comments below regarding not having enough logic/comments before classes and functions, but for a large team like this especially with some beginners it is imperative we try to make either better documentation in the code or in separate files.

Not only is it better to understand, it drills in some good software engineering practices that you should/will use in your career one way or another.

I focused mainly for classes and functions, however lots of nested/longer logic conditionals (if, elif, else, certain loops, etc.) should also have a blurb if the intention for them is not obvious at first glance.

File: src/moondawg/moondawg/serial_bridge.py [hunk 2]
```diff
@@ -21,49 +19,112 @@ def __init__(self):
...
+                            self.diag.level = DiagnosticStatus.ERROR
+                            self.diag.message = "Failed to recover from port bind failure."
+                            self.get_logger().error(self.diag.message + ": " + str(e))
+                            if self.serial_retry_timer.is_canceled():
+                                self.serial_retry_timer.reset()
+                        finally:
+                            return rclpy.parameter.SetParametersResult(successful=False)
+        return rclpy.parameter.SetParametersResult(successful=True)
+    
+    def connect_serial(self):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/serial_bridge.py:94]: added

File: src/moondawg/moondawg/serial_bridge.py [hunk 3]
```diff
@@ -21,49 +19,112 @@ def __init__(self):
...
             self.diag.message = "Serial connected."
-            self.serial_retry_timer.destroy()
-
+            self.serial_retry_timer.cancel()
+        except SerialException as e:
+            self.get_logger().error(f"Could not open connection to Arduino: {e}. Retrying in 5s")
+            self.diag.level = DiagnosticStatus.ERROR
+            self.diag.message = "Failure while trying to open serial connection."
 
     def serial_callback(self, message):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/serial_bridge.py:107]: added comment

File: src/moondawg/moondawg/serial_bridge.py [hunk 4]
```diff
@@ -21,49 +19,112 @@ def __init__(self):
...
+            self.get_logger().error(f"Error sending data to Arduino: {e}")
         except Exception as e:
-            self.get_logger().error("Error sending data to arduino: " + str(e))
+            self.get_logger().error(f"Unexpected error: {e}")
 
+    def stop(self):
+        if self.serial is not None:
+            self.serial.close()
             
     def heartbeat(self):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/serial_bridge.py:127]: added

File: src/moondawg/moondawg/serial_bridge.py [hunk 5]
```diff
@@ -21,49 +19,112 @@ def __init__(self):
...
         self.diag_topic.publish(self.diag)
+        
+        if self.get_logger().get_effective_level() <= rclpy.logging.LoggingSeverity.DEBUG:
+            self.get_logger().debug("Heartbeat published with diagnostic status: "
+                               f"Level: {self.diag.level}, Message: {self.diag.message}")
 
 def main(args=None):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/serial_bridge.py:134]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 6]
```diff
@@ -22,18 +23,21 @@
 
 class XboxTranslator(Node):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:24]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 7]
```diff
@@ -52,111 +56,60 @@ def __init__(self):
...
-        #     diff = (datetime.datetime.now() - self.time_start).total_seconds()
-        #     if (diff > 3 and diff < 5):
-        #         self.depositing = False
-        #         self.time_start = 0
-        #         message = self.deposit_string(0, forward)
-        #         self.serial_publisher.publish(message)
-        #     elif (diff > 5 and diff < 30):
-        #         pass
 
+    def auto_callback(self):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:88]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 8]
```diff
@@ -52,111 +56,60 @@ def __init__(self):
...
             self.get_logger().info("stopping now!!")
             self.time_start = 0
-            message = StringGen.belt_string(0)
-            self.serial_publisher.publish(message)
-            message = StringGen.belt_position_string(1, right)
-            self.serial_publisher.publish(message)
+            self.serial_publisher.publish(StringGen.belt_string(0))
+            self.serial_publisher.publish(StringGen.belt_position_string(1, right))
 
     def connection_callback(self, message):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:108]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 9]
```diff
@@ -52,111 +56,60 @@ def __init__(self):
...
-        self.connection_status = datetime.datetime.now()
+        self.connection_time = datetime.datetime.now()
         self.connected = 1
 
     def image_translator(self, message):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:113]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 10]
```diff
@@ -167,79 +120,62 @@ def parse_axis(self, data):
         }
     
     def calculate_speed(self, x, y):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:125]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 11]
```diff
@@ -167,79 +120,62 @@ def parse_axis(self, data):
...
 
         if (camera_pitch != self.camera_pitch):
-            self.serial_publisher.publish(self.camera_pitch_string(round(self.camera_pitch)))
             self.camera_pitch = camera_pitch
+            self.serial_publisher.publish(StringGen.camera_pitch_string(round(self.camera_pitch)))
 
-        
 
     # This function is called when the gamepad axis data is received
     def axis_callback(self, request):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py]: should the existing comment be clarified?

File: src/moondawg/moondawg/xbox_translator.py [hunk 12]
```diff
@@ -266,184 +202,164 @@ def parse_buttons(self, data):
             "dpad_right": data[15]/100,
         }
 
-    # This function is called when the gamepad button data is received
+    def check_auto(self, buttons):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:210]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 13]
```diff
@@ -266,184 +202,164 @@ def parse_buttons(self, data):
...
+        if (buttons['dpad_left']):
+            self.camera_arm = self.camera_arm + 5
+            self.camera_arm = max(0, min(self.camera_arm, 180))
+            self.serial_publisher.publish(StringGen.camera_arm_string(round(self.camera_arm)))
+        elif (buttons['dpad_right']):
+            self.camera_arm = self.camera_arm - 5
+            self.camera_arm = max(0, min(self.camera_arm, 180))
+            self.serial_publisher.publish(StringGen.camera_arm_string(round(self.camera_arm)))
+
     def button_callback(self, request):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:315]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 14]
```diff
@@ -266,184 +202,164 @@ def parse_buttons(self, data):
...
+            self.button_movement_handler(buttons)
+            self.misc_button_handler(buttons)
+
         except Exception as e:
             self.diag.level = DiagnosticStatus.WARN
             self.diag.message = "Exception in gamepad button callback."
             self.get_logger().error("Exception in gamepad button callback:" + str(e))
 
+    
+    def stop_all(self):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:336]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 15]
```diff
@@ -266,184 +202,164 @@ def parse_buttons(self, data):
...
+        self.serial_publisher.publish(StringGen.movement_string(90, 90))
+        self.serial_publisher.publish(StringGen.belt_speed_string(0, 90))
+        self.serial_publisher.publish(StringGen.belt_position_string(0, right))
+        self.serial_publisher.publish(StringGen.belt_string(0))
+        self.serial_publisher.publish(StringGen.deposit_string(0, forward))
+        self.serial_publisher.publish(StringGen.vibrator_string(0))
+
     def heartbeat(self):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:345]: added

File: src/moondawg/moondawg/xbox_translator.py [hunk 16]
```diff
@@ -266,184 +202,164 @@ def parse_buttons(self, data):
...
-        self.serial_publisher.publish(message)
-
-        message = StringGen.deposit_string(0, forward)
-        self.serial_publisher.publish(message)
-
-        message = StringGen.vibrator_string(0)
-        self.serial_publisher.publish(message)
 
 
 def main(args=None):
```
[the remark this answers is not in the document]
Comment (abarnes6) [reply src/moondawg/moondawg/xbox_translator.py:355]: added

**Pull request merged**